common/lib/common-utils/src/test/mocha/typedEventEmitter.spec.ts

```diff
@@ -48,4 +48,72 @@ describe("TypedEventEmitter", () => {
         assert.equal(newListenerCalls, 1);
         assert.equal(removeListenerCalls, 1);
     });
+
+    it("Validate emit typing for valid event type", () => {
+        const eventArgsEmitted: any[] = [];
+        const handler = (...args): void => {
+            eventArgsEmitted.push(args);
+        };
+        const tee = new (class SomeClass extends TypedEventEmitter<ISampleEvents> {
+            someMember: 0 = 0;
+        })();
+        const plainTee = new TypedEventEmitter();
+
+        tee.on("noArgs", handler);
+        tee.on("twoArgs", handler);
+        tee.on("useThis", handler);
+        tee.on("somethingElse", handler);
+
+        tee.emit("noArgs");
+        tee.emit("twoArgs", true, "hello");
+        tee.emit("useThis", tee);
+
+        // @ts-expect-error Unknown event
+        tee.emit("somethingElse");
+        // @ts-expect-error Unknown event
+        tee.emit("noArgs", "bogus");
+        // @ts-expect-error Wrong arg types
+        tee.emit("twoArgs", "wrongType", 123);
+        // @ts-expect-error Wrong arg type for "this"-typed arg
+        tee.emit("useThis", plainTee);
+
+        // @ts-expect-error This shouldn't be emitted manually
+        tee.emit("addListener", "asdf", () => {});
+        // @ts-expect-error This shouldn't be emitted manually
+        tee.emit("removeListener", "asdf", () => {});
+
+        // Even if typing is violated, events are still emitted as written
+        assert.deepStrictEqual(eventArgsEmitted, [
+            [],
+            [true, "hello"],
+            [tee],
+            [],
```

Comment:
After the first three lists of arguments here, do you care to assert what the rest are? Since they were only emitted by code that was a compile error it seems unnecessary to check them. You could move this assert up to line 70 to make the test a little shorter/simpler.

Reply:
I want to point out that if you override the typing somehow, it "works" - there's no runtime validation of the arguments.

Reply:
That's true. But should we write tests for cases when someone overrides the typing? What if I have a function:

```typescript
function getLength(str: string): number {
  return str.length;
}
```

Should I also write tests that check for correctness if I pass an `Array` into `getLength()`?
(That's not a completely fair analogy 😄; I'm fine with the test as is, just wanted to point out another option).